File: 152.maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if(nums.empty()) return 0;
        int ans = nums[0], max_ = nums[0], min_ = nums[0];
        for(int i = 1; i < nums.size(); i++) {
            int tmpMax = max_;
            int tmpMin = min_;
            max_ = max(max(tmpMax * nums[i], tmpMin * nums[i]), nums[i]);
            min_ = min(min(tmpMax * nums[i], tmpMin * nums[i]), nums[i]);
            ans = max(max_, ans);
        }
        return ans;
    }
};

File: fb/152.Maximum-Product-Subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        max_so_far = nums[0]
        min_so_far = nums[0]
        result = max_so_far
        
        for i in range(1, len(nums)):
            cur = nums[i]
            tmp_max = max(cur, max_so_far*cur, min_so_far*cur) // keep max_so_far unchange during max()/min() comparison
            min_so_far = min(cur, max_so_far*cur, min_so_far*cur)
            
            max_so_far = tmp_max
            result = max(max_so_far, result)
            
        return result
